scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+"""psrfits_translator.py defines the translator between baseband-style file
+object and psrfits file hdus.
+"""
+
+from .psrfits_fields import *
+from collections import namedtuple
+import astropy.units as u
+from astropy.time import Time
+from astropy.coordinates import SkyCoord
+from astropy.coordinates import Angle, Latitude, Longitude
+from astropy.io import fits
+import numpy as np
+
+
+__all__ = ["PsrfitsHearderHDU", "SubintHDU"]
+
+
+class HDUBase:
+    """This is the HDUBase class that defines some generic API functions
+    """
+    _header_defaults = {}
+    def verify(self):
+        raise NotImplementError
+
+    def set_header_card(self, key, value=None, comment=None):
+        if value is None and key in self._header_defaults.keys():
+            self.header.set(key, self._header_defaults['key']['value'],
+                            self._header_defaults['key']['comment'])
+        else:
+            self.header.set(key, value, comment)
+
+class PsrfitsHearderHDU(fits.PrimaryHDU, HDUBase):
+    """HeaderHDU class provides the translator function between baseband-style
+    file object and the PSRFITS main header HDU.
+
+    Parameter
+    ---------
+    header_hdu : `Pdat.header_hdu`
+    """
+    _properties = ('start_time', 'observatory', 'frequency', 'ra', 'dec',
+                   'shape', 'sample_rate')
+
+    _header_defaults = main_header
+
+    def __init__(self, header_hdu=None):
+        if header_hdu is None:
+            super(PsrfitsHearderHDU, self).__init__()
+            self._init_empty()
+        else:
+            super(PsrfitsHearderHDU, self).__init__(header=header_hdu.header,
+                                                    data=header_hdu.data)
+            self.verify()
+
+    def verify(self):
+        if 'SIMPLE' not in list(self.header.keys()):
+            raise ValueError("The input HDU is not a fits headers HDU.")
+        if self.header['FITSTYPE'] != "PSRFITS":
+            raise ValueError("The input fits header is not a PSRFITS type.")
+
+    def _init_empty(self):
+        for k, v in self._header_defaults.items():
+            self.header.set(k, v['value'], v['comment'])
+        self.header.add_comment('FITS (Flexible Image Transport System) format'
+                                'is defined in Astronomy and Astrophysics, '
+                                'volume 376, page 359; bibcode:'
+                                ' 2001A&A...376..359H', after='EXTEND')
+
+    @property
+    def start_time(self):
+        # TODO add location
+        MJD_d_int = np.longdouble(self.header['STT_IMJD']) * u.day
```

Comment:
Why is the `longdouble` needed?

Also, since `astropy` treats leap seconds differently than most pulsar timing experiments, we have to be a bit careful here. I think the following should be OK:
```
Time(self.header('STT_IMJD'], format='mjd') + TimeDelta(self.header['STT_SMJD'], self.header['STT_OFFS'], format='sec', scale='tai')
```